aptos-move/framework/src/natives/string_utils.rs

```diff
@@ -0,0 +1,233 @@
+        }
+        MoveTypeLayout::U16 => {
+            let u = val.value_as::<u16>()?;
+            write!(out, "{}", u).unwrap();
+        }
+        MoveTypeLayout::U32 => {
+            let u = val.value_as::<u32>()?;
+            write!(out, "{}", u).unwrap();
+        }
+        MoveTypeLayout::U256 => {
+            let u = val.value_as::<u256::U256>()?;
+            write!(out, "{}", u).unwrap();
+        }
+        MoveTypeLayout::Address => {
+            let addr = val.value_as::<move_core_types::account_address::AccountAddress>()?;
+            write!(out, "{}", addr).unwrap();
+        }
+        MoveTypeLayout::Signer => {
+            let signer = val.value_as::<move_core_types::account_address::AccountAddress>()?;
+            write!(out, "{}", signer).unwrap();
+        }
+        MoveTypeLayout::Vector(ty) => {
+            let v = val.value_as::<Vec<Value>>()?;
+            out.push('[');
+            for (i, x) in v.into_iter().enumerate() {
+                if i > 0 {
+                    out.push_str(", ");
+                }
+                native_format_impl(context, ty.as_ref(), x, out)?;
+            }
+            out.push(']');
+        }
+        MoveTypeLayout::Struct(MoveStructLayout::WithTypes { type_, fields, .. }) => {
+            let strct = val.value_as::<Struct>()?;
+            out.push_str(type_.name.as_str());
+            out.push_str(" {");
+            for (i, (ty, x)) in fields.iter().zip(strct.unpack()?).enumerate() {
+                if i > 0 {
+                    out.push_str(", ");
+                }
+                write!(out, "{}: ", ty.name).unwrap();
+                native_format_impl(context, &ty.layout, x, out)?;
+            }
+            out.push('}');
+        }
+        MoveTypeLayout::Struct(MoveStructLayout::WithFields(fields)) => {
+            let strct = val.value_as::<Struct>()?;
+            out.push('{');
+            for (i, (ty, x)) in fields.iter().zip(strct.unpack()?).enumerate() {
+                if i > 0 {
+                    out.push_str(", ");
+                }
+                write!(out, "{}: ", ty.name).unwrap();
+                native_format_impl(context, &ty.layout, x, out)?;
+            }
+            out.push('}');
+        }
+        MoveTypeLayout::Struct(MoveStructLayout::Runtime(fields)) => {
+            let strct = val.value_as::<Struct>()?;
+            out.push('(');
+            for (i, (ty, x)) in fields.iter().zip(strct.unpack()?).enumerate() {
+                if i > 0 {
+                    out.push_str(", ");
+                }
+                native_format_impl(context, ty, x, out)?;
+            }
+            out.push(')');
+        }
+    };
+    Ok(())
+}
+
+pub fn native_format(
+    _gas_params: &FormatGasParams,
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut arguments: VecDeque<Value>,
+    ) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    debug_assert!(ty_args.len() == 1);
```

Comment:
We need to charge gas appropriately for all these new native functions and add those gas params to the gas schedule https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/aptos-gas/src/aptos_framework.rs

Reply:
I added gas